### PR TITLE
feat: support custom amp cast list

### PR DESCRIPTION
**Pull request by The-truthh**
Thank you for your contribution to the MindCV repo.
Before submitting this PR, please make sure:

- [x] You have read the [Contributing Guidelines on pull requests](https://github.com/mindspore-lab/mindcv/blob/main/CONTRIBUTING.md)
- [x] Your code builds clean without any errors or warnings
- [x] You are using approved terminology
- [ ] You have added unit tests

## Motivation

support custom amp cast list

## Test Plan

(How should this PR be tested? Do you require special setup to run the test or repro the fixed bug?)

## Related Issues and PRs

(Is this PR part of a group of changes? Link the other relevant PRs and Issues here. Use https://help.github.com/en/articles/closing-issues-using-keywords for help on GitHub syntax)

### CODE REVIEW

File: config.py [hunk 1]
```diff
@@ -232,6 +232,12 @@ def create_parser():
                             'Choice: O0 - all FP32, O1 - only cast ops in white-list to FP16, '
                             'O2 - cast all ops except for blacklist to FP16, '
                             'O3 - cast all ops to FP16. (default="O0").')
+    group.add_argument('--amp_cast_list', type=str, default=None,
```
Comment (vigo999) [inline config.py:235]: can rewrite this help msg? can't be easily understand

**Comment by vigo999** [comment]
what is the difference with original api below?
[mindcv/utils/amp.py](https://www.mindspore.cn/docs/en/r2.0/_modules/mindspore/train/amp.html#auto_mixed_precision)

**Comment by The-truthh** [comment]
The difference of function 'auto_mixed_precision' between mindspore and mindcv is that mindcv supports custom black list and white list. If we use the default black list and white list, the effect is the same.

File: mindcv/utils/trainer_factory.py [hunk 2]
```diff
@@ -84,13 +113,18 @@ def create_trainer(
     if gradient_accumulation_steps < 1:
         raise ValueError("`gradient_accumulation_steps` must be >= 1!")
 
+    # process amp with custom cast list
```
Comment (geniuspatrick) [inline mindcv/utils/trainer_factory.py:116]: 先只对自定义训练逻辑实现自定义黑白名单

File: mindcv/utils/trainer_factory.py [hunk 3]
```diff
@@ -84,13 +113,18 @@ def create_trainer(
...
+    eval_network = nn.WithEvalCell(network, loss, amp_level in ["O2", "O3", "auto"])
+    auto_mixed_precision(network, amp_level, amp_cast_list=amp_cast_list)
+    net_with_loss = add_loss_network(network, loss, amp_level)
+
     if not require_customized_train_step(ema, clip_grad, gradient_accumulation_steps):
```
Comment (geniuspatrick) [inline mindcv/utils/trainer_factory.py:121]: 这个代码块下的逻辑不要改变，保持绝对的兼容性

File: mindcv/utils/trainer_factory.py [hunk 4]
```diff
@@ -107,12 +141,12 @@ def create_trainer(
                 "when creating a mindspore.Model instance. "
                 "NOTE: mindspore.Model may use LossScaleManager silently. See mindspore.train.amp for details."
             )
+            if amp_level == "O2":
+                mindspore_kwargs["loss_scale_manager"] = DynamicLossScaleManager()
         else:
             raise ValueError(f"Loss scale type only support ['fixed', 'dynamic', 'auto'], but got{loss_scale_type}.")
         model = Model(**mindspore_kwargs)
     else:  # require customized train step
```
Comment (geniuspatrick) [inline mindcv/utils/trainer_factory.py:149]: 对这个分支，增添自定义名单逻辑

File: mindcv/utils/amp.py [hunk 5]
```diff
@@ -61,3 +64,130 @@ def unscale(self, inputs):
 
     def adjust(self, grads_finite):
         return True
+
+
+AMP_WHITE_LIST = (
```
Comment (geniuspatrick) [inline mindcv/utils/amp.py:69]: 有些函数和变量可以从mindspore导入的，直接导入；除非我们需要覆盖或重载函数

**Review by geniuspatrick — CHANGES_REQUESTED** [review]

**Review by SamitHuang — APPROVED** [review]
Update training tutorial for this new feature of amp customization.

**Review by SamitHuang — APPROVED** [review]
Update training tutorial for this new feature of amp customization.

File: config.py [hunk 6]
```diff
@@ -237,6 +237,11 @@ def create_parser():
...
+                       help='At the cell level, customize the list to cast ops to FP16. '
+                            'For O1, cast all ops in the list to FP16, '
+                            'For O2, cast all ops except for the list to FP16. '
```
Comment (geniuspatrick) [inline config.py:243]: ex1:  amp_cast_list: "['nn.Conv2D']"

**Review by geniuspatrick — APPROVED** [review]

File: config.py (hunk 6 above)
Comment (geniuspatrick) [reply config.py:243]: examples & explanation

**Review by geniuspatrick — CHANGES_REQUESTED** [review]

**Review by geniuspatrick — APPROVED** [review]

**Pull request merged**